src/psyclone/tests/psyir/backend/fortran_gen_decls_test.py

```diff
@@ -265,3 +266,18 @@ def test_gen_decls_routine_wrong_interface(fortran_writer):
     assert ("Routine symbol 'arg_sub' is passed as an argument (has an "
             "ArgumentInterface). This is not supported by the Fortran "
             "back-end." in str(info.value))
+
+
+def test_gen_decls_static_variables(fortran_writer):
+    '''Test that the gen_decls and gen_vardecl methods add the appropriate
+    Fortran attributes to static variables.
+
+    '''
+    symbol_table = SymbolTable()
+    # Add a routine symbol with an unsupported ArgumentInterface
+    sym = DataSymbol("v1", datatype=INTEGER_TYPE, interface=StaticInterface())
+    symbol_table.add(sym)
+    assert "integer, save :: v1" in fortran_writer.gen_decls(symbol_table)
+    assert "integer, save :: v1" in fortran_writer.gen_vardecl(sym)
+    sym.constant_value = 1
+    assert "parameter, save :: v1 = 1" in fortran_writer.gen_vardecl(sym)
```

Comment:
gfortran does not accept a `save` attribute for a parameter. Since a parameter is a compile-time constant I guess it doesn't make sense to declare it as static:

        3 |   integer, parameter, save :: n = 10
          |                          1
    Error: PARAMETER attribute conflicts with SAVE attribute at (1)

Reply:
This has let me to discover that we were happily accepting a lot of attribute combinations that are not valid Fortran: e.g. allocate with parameter, save with intent,  intents with other intents, ... I have added a bunch of new checks in the fparser frontend to capture this and produce errors early.

However, for the PSyIR the save and has_constant doesn't seem harmful because a parameter already behaves like a static variable. So I just made sure that the backend only outputs the parameter attribute and does not generate invalid code.